source/fab/steps/analyse.py

```diff
@@ -0,0 +1,352 @@
+        self.unreferenced_deps: List[str] = unreferenced_deps or []
+
+        # todo: these seem more like functions
+        self.fortran_analyser = FortranAnalyser(std=std)
+        self.c_analyser = CAnalyser()
+
+    def run(self, artefacts, config):
+        """
+        Creates the *build_tree* artefact: Dict[Path, AnalysedFile] from the files in `self.source_getter`.
+
+        Steps, in order:
+            - Creates a hash of every artefact, used to check if it's already been analysed.
+            - Parse the C and Fortran files to find the symbol definitions and depndencies in each file.
+                - Analysis results are stored in a csv as-we-go, so analysis can be resumed if interrupted.
+            - Creates a symbol table for external symbols: Dict[symbol, Path]
+            - Work out the file dependencies from the symbol dependencies.
+            - (Optionally) Prune the file dependency tree for the root symbol, if given.
+
+        This step uses multiprocessing, unless disabled in the :class:`~fab.steps.Step` class.
+
+        """
+        super().run(artefacts, config)
+
+        files = self.source_getter(artefacts)
+
+        # take hashes of all the files we preprocessed
+        with time_logger(f"getting {len(files)} hashes"):
+            preprocessed_hashes = self._get_latest_checksums(files)
+
+        with time_logger("loading previous analysis results"):
+            changed, unchanged = self._load_analysis_results(preprocessed_hashes)
+
+        with time_logger("analysing files"):
+            with self._new_analysis_file(unchanged) as csv_writer:
+                analysed_fortran, analysed_c = self._parse_files(changed, csv_writer)
+        all_analysed_files: Dict[Path, AnalysedFile] = {a.fpath: a for a in unchanged + analysed_fortran + analysed_c}
+
+        # Make "external" symbol table
+        with time_logger("creating symbol lookup"):
+            symbols: Dict[str, Path] = self._gen_symbol_table(all_analysed_files)
+
+        # turn symbol deps into file deps
+        with time_logger("generating file dependencies from symbols"):
+            self._gen_file_deps(all_analysed_files, symbols)
+
+        #  find the file dependencies for UM "DEPENDS ON:" commented file deps
+        with time_logger("adding MO 'DEPENDS ON:' file dependency comments"):
+            add_mo_commented_file_deps(analysed_fortran, analysed_c)
+
+        logger.info(f"source tree size {len(all_analysed_files)}")
+
+        # Target tree extraction - for building executables.
+        if self.root_symbol:
+            with time_logger("extracting target tree"):
+                build_tree = extract_sub_tree(all_analysed_files, symbols[self.root_symbol], verbose=False)
+            logger.info(f"build tree size {len(build_tree)} (target '{symbols[self.root_symbol]}')")
+        # When building library ".so" files, no target is needed.
+        else:
```

Comment:
We must ensure the current implementation is flexible enough that it can be modified to cater for new requirements with minimal effort, and result in a clean new design. It's hard to comment on this specific suggestion without concrete requirements but I'm sure we're all confident that we'll be able to implement it to a high standard if the need arises. It's worth pointing out that we're already building two versions of the gcom library from a single source grab.